### PR TITLE
ORC-694: Update docker files for adding Java11 support

### PR DESCRIPTION
**Pull request by yuokada**
### What changes were proposed in this pull request?

- Add jdk argument in Dockerfile to support JDK11 

### Why are the changes needed?

- Since github actions for the repository use java 11, we should provide the images using JDK 11

### How was this patch tested?

- Confirmed that `Dockerfile`s are built successfully with `reinit.sh`
- `bash -xe ./run-one.sh yuokada ORC-694 ubuntu20`

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for working on this, @yuokada .

BTW, do we need to change `MAINTAINER` in this PR?

**Comment by yuokada** [comment]
@dongjoon-hyun 
`MAINTAINER` command is deprecated since docker 1.13 or later. So I replaced it with the `LABEL` command.
https://docs.docker.com/engine/reference/builder/#maintainer-deprecated

But, we don't need to change it in this PR.

Should we create another ticket to replace `MAINTAINER` command?

**Comment by dongjoon-hyun** [comment]
Shall we do that in a separate PR because it's irrelevant to Java?

**Comment by yuokada** [comment]
Created: https://github.com/apache/orc/pull/656

File: docker/README.md [hunk 1]
```diff
@@ -21,7 +21,7 @@ The scripts are:
 A base image for each OS is built using:
 
     cd docker/$os
-    docker build -t orc-$os .
+    docker build -t "orc-$os-jdk8" --build-arg jdk=8 . OR docker build -t "orc-$os-jdk11" --build-arg jdk=11 .
```
Comment (yuokada) [inline docker/README.md:24]: Is this change easy to understand?

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Could you minimize the size of patch at `CentOS Dockerfile`? Technically, `if` statement should be related to `java-xx-openjdk`. For the other part `git/maven/make/...`, we don't need to duplicate.

File: docker/ubuntu18/Dockerfile [hunk 2]
```diff
@@ -32,9 +33,9 @@ RUN apt-get install -y \
   make \
   curl \
   maven \
-  openjdk-8-jdk \
+  openjdk-${jdk}-jdk \
   tzdata
-RUN update-java-alternatives --set java-1.8.0-openjdk-amd64
+RUN update-java-alternatives --set java-1.${jdk}.0-openjdk-amd64
```
Comment (dongjoon-hyun) [inline docker/ubuntu18/Dockerfile:38]: Ur, is this correct, `java-1.11.0-openjdk-amd64`?

File: docker/ubuntu20-clang/Dockerfile [hunk 3]
```diff
@@ -31,9 +32,9 @@ RUN apt-get install -y \
   make \
   curl \
   maven \
-  openjdk-8-jdk \
+  openjdk-${jdk}-jdk \
   tzdata
-RUN update-java-alternatives --set java-1.8.0-openjdk-amd64
+RUN update-java-alternatives --set java-1.${jdk}.0-openjdk-amd64
```
Comment (dongjoon-hyun) [inline docker/ubuntu20-clang/Dockerfile:37]: ditto.

File: docker/ubuntu20/Dockerfile [hunk 4]
```diff
@@ -32,9 +33,9 @@ RUN apt-get install -y \
   make \
   curl \
   maven \
-  openjdk-8-jdk \
+  openjdk-${jdk}-jdk \
   tzdata
-RUN update-java-alternatives --set java-1.8.0-openjdk-amd64
+RUN update-java-alternatives --set java-1.${jdk}.0-openjdk-amd64
```
Comment (dongjoon-hyun) [inline docker/ubuntu20/Dockerfile:38]: ditto.

File: docker/ubuntu18/Dockerfile (hunk 2 above)
Comment (yuokada) [reply docker/ubuntu18/Dockerfile:38]: I believe so. It properly is set.
```
~/w/I/o/docker ❯❯❯ docker run -it orc-ubuntu18-jdk11 update-java-alternatives -l   
java-1.11.0-openjdk-amd64      1111       /usr/lib/jvm/java-1.11.0-openjdk-amd64

~/w/I/o/docker ❯❯❯ docker run -it orc-ubuntu18-jdk11 java -version  
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment (build 11.0.10+9-Ubuntu-0ubuntu1.18.04)
OpenJDK 64-Bit Server VM (build 11.0.10+9-Ubuntu-0ubuntu1.18.04, mixed mode, sharing)
```

File: docker/ubuntu20/Dockerfile (hunk 4 above)
Comment (yuokada) [reply docker/ubuntu20/Dockerfile:38]:
```
~/w/I/o/docker ❯❯❯ docker run -it orc-ubuntu20-jdk11 update-java-alternatives -l  
java-1.11.0-openjdk-amd64      1111       /usr/lib/jvm/java-1.11.0-openjdk-amd64

~/w/I/o/docker ❯❯❯ docker run -it orc-ubuntu20-jdk11 java -version
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment (build 11.0.10+9-Ubuntu-0ubuntu1.20.04)
OpenJDK 64-Bit Server VM (build 11.0.10+9-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)
```

File: docker/run-all.sh [hunk 5]
```diff
@@ -32,9 +32,11 @@ function failure {
 rm -f logs/pids.txt logs/*.log
 
 start=`date`
-for os in `cat os-list.txt`; do
-  echo "Building $os"
-  ( cd $os && docker build -t "orc-$os" . ) > logs/$os-build.log 2>&1 || exit 1
+for jdk in 8 11; do
+    for os in `cat os-list.txt`; do
+        echo "Building $os"
```
Comment (dongjoon-hyun) [inline docker/run-all.sh:37]: Could you show `$jdk` together?

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for updating, @yuokada . The PR looks good mostly. I left one minor comment about log message.

File: docker/README.md (hunk 1 above)
Comment (pgaref) [reply docker/README.md:24]: Lets make it more explicit.
For jdk8:  docker build -t "orc-$os-jdk8" --build-arg jdk=8 .
For jdk11: docker build -t "orc-$os-jdk11" --build-arg jdk=11 .

File: docker/README.md [hunk 6]
```diff
@@ -21,7 +21,7 @@ The scripts are:
 A base image for each OS is built using:
```
Comment (pgaref) [inline docker/README.md:21]: Lets add above that:
`run-all.sh` and `run-one.sh` tests both on jdk8 and 11 across OSes

**Comment by pgaref** [comment]
Thanks @yuokada  and @dongjoon-hyun  for the work on this! Left some minor comments.
Seems to work well locally!

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I built and ran this PR. 
```bash
$ for i in $(docker images | grep orc | awk '{print $1}'); do echo $i; docker run -it --rm $i java -version; done
orc-ubuntu20-clang-jdk11
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment (build 11.0.10+9-Ubuntu-0ubuntu1.20.04)
OpenJDK 64-Bit Server VM (build 11.0.10+9-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)
orc-ubuntu20-jdk11
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment (build 11.0.10+9-Ubuntu-0ubuntu1.20.04)
OpenJDK 64-Bit Server VM (build 11.0.10+9-Ubuntu-0ubuntu1.20.04, mixed mode, sharing)
orc-ubuntu18-jdk11
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment (build 11.0.10+9-Ubuntu-0ubuntu1.18.04)
OpenJDK 64-Bit Server VM (build 11.0.10+9-Ubuntu-0ubuntu1.18.04, mixed mode, sharing)
orc-centos8-jdk11
openjdk version "11.0.9.1" 2020-11-04 LTS
OpenJDK Runtime Environment 18.9 (build 11.0.9.1+1-LTS)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.9.1+1-LTS, mixed mode, sharing)
orc-centos7-jdk11
openjdk version "11.0.10" 2021-01-19 LTS
OpenJDK Runtime Environment 18.9 (build 11.0.10+9-LTS)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.10+9-LTS, mixed mode, sharing)
orc-ubuntu20-clang-jdk8
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (build 1.8.0_282-8u282-b08-0ubuntu1~20.04-b08)
OpenJDK 64-Bit Server VM (build 25.282-b08, mixed mode)
orc-ubuntu20-jdk8
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (build 1.8.0_282-8u282-b08-0ubuntu1~20.04-b08)
OpenJDK 64-Bit Server VM (build 25.282-b08, mixed mode)
orc-ubuntu18-jdk8
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (build 1.8.0_282-8u282-b08-0ubuntu1~18.04-b08)
OpenJDK 64-Bit Server VM (build 25.282-b08, mixed mode)
orc-ubuntu16-jdk11
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (build 1.8.0_282-8u282-b08-0ubuntu1~16.04-b08)
OpenJDK 64-Bit Server VM (build 25.282-b08, mixed mode)
orc-ubuntu16-jdk8
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (build 1.8.0_282-8u282-b08-0ubuntu1~16.04-b08)
OpenJDK 64-Bit Server VM (build 25.282-b08, mixed mode)
orc-debian10-jdk11
openjdk version "11.0.9.1" 2020-11-04
OpenJDK Runtime Environment (build 11.0.9.1+1-post-Debian-1deb10u2)
OpenJDK 64-Bit Server VM (build 11.0.9.1+1-post-Debian-1deb10u2, mixed mode, sharing)
orc-debian10-jdk8
openjdk version "11.0.9.1" 2020-11-04
OpenJDK Runtime Environment (build 11.0.9.1+1-post-Debian-1deb10u2)
OpenJDK 64-Bit Server VM (build 11.0.9.1+1-post-Debian-1deb10u2, mixed mode, sharing)
orc-debian9-jdk11
openjdk version "1.8.0_275"
OpenJDK Runtime Environment (build 1.8.0_275-8u275-b01-1~deb9u1-b01)
OpenJDK 64-Bit Server VM (build 25.275-b01, mixed mode)
orc-debian9-jdk8
openjdk version "1.8.0_275"
OpenJDK Runtime Environment (build 1.8.0_275-8u275-b01-1~deb9u1-b01)
OpenJDK 64-Bit Server VM (build 25.275-b01, mixed mode)
orc-centos8-jdk8
openjdk version "1.8.0_275"
OpenJDK Runtime Environment (build 1.8.0_275-b01)
OpenJDK 64-Bit Server VM (build 25.275-b01, mixed mode)
orc-centos7-jdk8
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (build 1.8.0_282-b08)
OpenJDK 64-Bit Server VM (build 25.282-b08, mixed mode)
```

In short, the following three look suspicious to me. Could you double-check it, @yuokada ?
```
orc-ubuntu16-jdk11
openjdk version "1.8.0_282"
OpenJDK Runtime Environment (build 1.8.0_282-8u282-b08-0ubuntu1~16.04-b08)
OpenJDK 64-Bit Server VM (build 25.282-b08, mixed mode)
...
orc-debian10-jdk8
openjdk version "11.0.9.1" 2020-11-04
OpenJDK Runtime Environment (build 11.0.9.1+1-post-Debian-1deb10u2)
OpenJDK 64-Bit Server VM (build 11.0.9.1+1-post-Debian-1deb10u2, mixed mode, sharing)
...
orc-debian9-jdk11
openjdk version "1.8.0_275"
OpenJDK Runtime Environment (build 1.8.0_275-8u275-b01-1~deb9u1-b01)
OpenJDK 64-Bit Server VM (build 25.275-b01, mixed mode)
```

**Comment by yuokada** [comment]
@dongjoon-hyun 
Just questions.
For debian 10, the OS image already use jdk10. Should we support jdk8 on it?

For debian 9 and ubuntu16, the openjdk-11 deb package isn't provided for both OSes.
Do we need to support JDK 11 in both?

**Comment by dongjoon-hyun** [comment]
If there is no JDK11 package, we had better document it and skip JDK11 for old OSes inside the above script like run-all.sh.

**Comment by dongjoon-hyun** [comment]
Thank you, @yuokada and @omalley .

**Pull request closed without merging**